pkg/operator/configobserver/cloudprovider/observe_cloudprovider.go

```diff
@@ -66,7 +68,25 @@ func (c *cloudProviderObserver) ObserveCloudProviderNames(genericListers configo
 		return existingConfig, append(errs, err)
 	}
 
+	featureGate, err := listers.FeatureGateLister().Get("cluster")
+	if errors.IsNotFound(err) {
+		recorder.Eventf("ObserveCloudProviderNames", "Optional featuregate.%s/cluster not found", configv1.GroupName)
+	} else if err != nil {
+		return existingConfig, append(errs, err)
+	}
+
+	external, err := cloudprovider.IsCloudProviderExternal(infrastructure.Status.Platform, featureGate)
+	if err != nil {
+		recorder.Eventf("ObserveCloudProviderNames", "Could not determine external cloud provider state: %v", err)
+	} else if external {
+		if err := unstructured.SetNestedStringSlice(observedConfig, []string{"external"}, c.cloudProviderNamePath...); err != nil {
+			errs = append(errs, err)
+		}
+		return observedConfig, errs
```

Comment:
You need to set `if err := listers.ResourceSyncer().SyncConfigMap` to the correct (empty I think) value, in this case.

Reply:
That is right, but https://github.com/openshift/library-go/pull/994 requires this cloud-config to be copied even for `external` cloud-config option, as there still will be a consumer in some cases, like for KCM.

Reply:
> That is right, but #994 requires this cloud-config to be copied even for `external` cloud-config option, as there still will be a consumer in some cases, like for KCM.

Coding by side-effect is bad practice. As I see it, if you want to rely on the same configmap, it makes sense to produce a separate configobserver function which both of these methods can rely upon.

As coded here, you aren't doing any copy when you start external anyway, so this broke the active resource sync.

Reply:
You are right, sorry for confusion. We definitely don't need a side effect on configobserver which will copy cloud-config to destination for the --external-cloud-provider-plugin flag in another PR, it will do this separately. But we should not remove it from there as well, so I think that this part is currently correct. It won't remove cloud-config which some other observer might place there. https://github.com/openshift/library-go/pull/994 will address that.